wave_utils/include/wave/utils/pose_cov_comp.hpp

```diff
@@ -0,0 +1,165 @@
+ *           and on-manifold optimization,” Univ. Malaga, Tech. Rep, no. 3,
+ *           pp. 1–56, 2014.
+ * Auth: Chunshang Li and Jordan Hu
+ *
+ * ############################################################################
+ */
+
+#ifndef REF_EKF_POSE_COV_COMP_HPP_
+#define REF_EKF_POSE_COV_COMP_HPP_
+
+#include <Eigen/Core>
+#include <Eigen/Dense>
+
+namespace wave {
+
+typedef Eigen::Matrix<double, 7, 1> Vector7;
+typedef Eigen::Matrix<double, 6, 1> Vector6;
+typedef Eigen::Matrix<double, 4, 1> Vector4;
+typedef Eigen::Matrix<double, 3, 1> Vector3;
+typedef Eigen::Matrix<double, 6, 6, Eigen::RowMajor> Matrix6x6;
+typedef Eigen::Matrix<double, 6, 7, Eigen::RowMajor> Matrix6x7;
+typedef Eigen::Matrix<double, 7, 6, Eigen::RowMajor> Matrix7x6;
+typedef Eigen::Matrix<double, 4, 4, Eigen::RowMajor> Matrix4x4;
+typedef Eigen::Matrix<double, 3, 4, Eigen::RowMajor> Matrix3x4;
+typedef Eigen::Matrix<double, 7, 7, Eigen::RowMajor> Matrix7x7;
+typedef Eigen::Matrix<double, 3, 7, Eigen::RowMajor> Matrix3x7;
+typedef Eigen::Matrix<double, 3, 3, Eigen::RowMajor> Matrix3x3;
+
+struct PoseWithCovariance {
+    Vector3 position;
+    Matrix3x3 rotation_matrix;
+    Matrix6x6 covariance;
+
+    PoseWithCovariance();
+    PoseWithCovariance(Vector6 &p, Matrix6x6 &q);
+    PoseWithCovariance(Vector3 &p, Matrix3x3 &r, Matrix6x6 &q);
+    Vector3 getPosition() const;
+    Vector3 getYPR() const;
+    Vector4 getQuaternion() const;
+    Vector7 getPoseQuaternion() const;
+    Eigen::Affine3d getTransformMatrix() const;
+};
+
+/*
+ * Pose composition based on [1]
+ */
+PoseWithCovariance composePose(PoseWithCovariance &p1, PoseWithCovariance &p);
+
+/*
+ * Normalizes Quaternion where qr^2 + qx^2 + qy^2 + qz^2 = 1
+ * Equation(1.6)
+ * @return: Vector4 [qw, qx, qy, qz]
+ */
+Vector4 normalizeQuat(const Vector4 q);
+
+/*
+ * Converts Quaternion to Yaw Pitch Roll angles
+ * Equation(2.9) and Equation(2.10)
+ * Note v(0) is yaw, v(1) is pitch, v(2) is roll
+ * @return: Vector4 [qw, qx, qy, qz]
+ */
+Vector3 quatToYPR(const Vector4 &q);
+
+/*
+ * Converts Yaw Pitch Roll angles to Quaternion
+ * Equation(2.3), Equation(2.4), Equation(2.5), Equation(2.6)
+ * @return: Vector3 [yaw, pitch, roll]
+ */
+Vector4 yprToQuat(const Vector3 &ypr);
+
+/*
+ * Converts Yaw Pitch Roll angles to Rotation Matrix
+ * Equation(2.18)
+ * @return: Matrix3x3
+ */
+Matrix3x3 yprToRotMatrix(const Vector3 &ypr);
+
+/*
+ * Converts Rotation Matrix to Yaw Pitch Roll angles
+ * Equation(2.20), Equation(2.21), Equation(2.22), and Equation(2.23)
+ * @return: Vector3 [yaw, pitch, roll]
+ */
+Vector3 rotMatrixToYPR(const Matrix3x3 &p);
+
+/*
+ * Converts Rotation Matrix to Quaternion
+ * According to [1], fastest way to convert from Rotation Matrix to Quat is to
+ * convert Quaternion -> Euler Angles -> Rotation Matrix
+ * @return: Vector4 [qw, qx, qy, qz]
+ */
+Vector4 rotMatrixToQuat(const Matrix3x3 &p);
+
+/// the jacobian of quaternion normalization function
```

Comment:
Not Doxygen

Reply:
is `///` or `/** */` preferred in general?

Reply:
Converted all to `/** */` format. Done

Reply:
Oh whoops I didn't realize there were 3 ///. My bad! Thanks for changing it.